### ISSUE TITLE
Rate limiting handle

### ISSUE THREAD

**Issue by atassis** [telegraf/telegraf#1207]
**Is your feature request related to a problem? Please describe.**
Telegram is rate limiting requests. The telegraf itselves is not handling this case

**Describe the solution you'd like**
To create an inner mechanism for handling API calls (both rate limiter and 429 retry mechanisms)

**Comment by wojpawlik** [comment]
See https://github.com/KnightNiwrem/telegraf-throttler

**Comment by KnorpelSenf** [comment]
And https://github.com/telegraf/telegraf-ratelimit

**Comment by atassis** [comment]
@wojpawlik that's what I was talking about. Would be better if telegraf closed this case by itself since it's a framework and this is an internal behavior (and the module changes an API call).

**Comment by wojpawlik** [comment]
`Telegram::callApi` is actually part of public interface, so it's safe to replace with a method with the same signature.

Ratelimiting would introduce delays into handling updates, which is a problem: #892.

**Comment by KnorpelSenf** [comment]
May be reconsidered if #1234 is done well

**Comment by wojpawlik** [comment]
I don't think #1234 helps much. It's safe to poll for further updates only after current updates are finished processing, because they won't be sent again. It's not safe to poll for further updates when an update is waiting for rate limits to finish processing. Until it's possible outside webhooks, I doubt rate-limiting support can be brought into core.

**Comment by atassis** [comment]
@wojpawlik I still think that framework needs to handle this case by itself because it is a pretty common issue, which needs the developer much deeper research of telegram bot API. If a framework makes its user searching this king of things too many times- it is becoming less "frameworky"

**Comment by KnorpelSenf** [comment]
I mean technically we could easily just merge that external project with our code base and let the user opt-in by setting a flag. Question is a) if we even _want to have_ automatic handling of rate limiting in `telegraf` itself, and b) if we do, whether there is a better solution that rewriting methods.

We can easily dismiss this as won't fix if we deem it out of scope. You wouldn't want to have that in large bots, you'd definitely want to be able to handle 429s yourself. Rate limiting is usually not a temporary issue but a problem of too high load, which requires some re-engineering. Quite the contrary, many people would indeed like to have that in small bots with few users because then it is really probably temporary.

**Comment by wojpawlik** [comment]
I understand your stance @atassis, however:

1. Let's not argue about semantics (framework / library),
2. I value reliability more than convenience, users can easily trade the former for the latter but not vice versa,
3. I would rather avoid introducing a [leaky abstraction](https://www.joelonsoftware.com/2002/11/11/the-law-of-leaky-abstractions/) into Telegraf,
4. If rate limit handling should be optional, why force users who won't use it to download the code for it and its dependencies?

**Comment by wojpawlik** [comment]
We're thinking how to allow 3rd-party rate limiting without replacing `callApi`: https://github.com/telegraf/telegraf/discussions/1267.

**Comment by atassis** [comment]
Cool, that's a thing!

**Comment by 2naive** [comment]
@wojpawlik so are there any news on 429 retries?
